Compiler problem after upgrading the Nightly channel to 1.26.0-nightly (521d91c6b 2018-03-14), on x86_64-apple-darwin

# Action Done
Upgrading nightly channel (on Mac OS, x86_64).

# Error

error: internal compiler error: librustc/ich/impls_ty.rs:907: ty::TypeVariants::hash_stable() - Unexpected variant TyInfer(?0).

# Environment
rustc 1.26.0-nightly (521d91c6b 2018-03-14) running on x86_64-apple-darwin
compiler flags: -C debuginfo=2 -C incremental --crate-type lib

<details>
<summary>Full Trace</summary>

```
thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:540:9
stack backtrace:
   0:        0x10ae4ccd3 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h4f12297ed49ac32f
   1:        0x10ae43dd2 - std::sys_common::backtrace::_print::h213e0e54268e0593
   2:        0x10ae496cd - std::panicking::default_hook::{{closure}}::ha2d6c27d3ccb45fe
   3:        0x10ae49358 - std::panicking::default_hook::h05e34c08d7c0020c
   4:        0x10943ca40 - core::ops::function::Fn::call::h33445f89d9a18bf0
   5:        0x10ae49b90 - std::panicking::rust_panic_with_hook::h786d5042983c8bdd
   6:        0x10abdd33d - std::panicking::begin_panic::hbc36fe3fd409c901
   7:        0x10abfe6f8 - rustc_errors::Handler::bug::h6b903b69d6f4723b
   8:        0x10975d988 - rustc::session::opt_span_bug_fmt::{{closure}}::h7c0a304ef8c6532d
   9:        0x10981f2c0 - rustc::ty::context::tls::with_opt::{{closure}}::h3d2d9ce484fff7ea
  10:        0x109433d68 - <std::thread::local::LocalKey<T>>::try_with::h3c186cf53b87e91f
  11:        0x1094339c4 - <std::thread::local::LocalKey<T>>::with::h1e00bc547b6c14e5
  12:        0x10981f104 - rustc::ty::context::tls::with::h22abd2a67eff726d
  13:        0x10981f27f - rustc::ty::context::tls::with_opt::h2ae30e76357ed40a
  14:        0x10975d894 - rustc::session::opt_span_bug_fmt::h6e5624c321dff564
  15:        0x10975d7e1 - rustc::session::bug_fmt::hfd486ef4eb0c27f2
  16:        0x109690720 - rustc::ich::impls_ty::<impl rustc_data_structures::stable_hasher::HashStable<rustc::ich::hcx::StableHashingContext<'a>> for rustc::ty::sty::TypeVariants<'gcx>>::hash_stable::h0e6587b9a6d0d5ad
  17:        0x109841cbb - rustc::dep_graph::dep_node::DepNode::new::h4e54c5801157ea48
  18:        0x1099e796f - rustc::ty::maps::<impl rustc::ty::maps::queries::dropck_outlives<'tcx>>::try_get::h078dfc868290b2fa
  19:        0x1099fd6be - rustc::ty::maps::TyCtxtAt::dropck_outlives::h88b104672b4b7329
  20:        0x1097ac3c9 - rustc::traits::query::dropck_outlives::<impl rustc::infer::at::At<'cx, 'gcx, 'tcx>>::dropck_outlives::h66306048d6ba5805
  21:        0x1086b5eed - rustc_typeck::check::dropck::check_safety_of_destructor_if_necessary::h4e601f460f34e015
  22:        0x1086c24f3 - rustc_typeck::check::regionck::RegionCtxt::check_safety_of_rvalue_destructor_if_necessary::h99ebdbedfe415d3f
  23:        0x1086c077a - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  24:        0x10868070e - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  25:        0x1086c0f74 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  26:        0x1086c0439 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_local::hfb7a8d29d91624dd
  27:        0x108680aa5 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  28:        0x1086c0e57 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  29:        0x108680962 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  30:        0x1086c112b - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  31:        0x108680ac0 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  32:        0x1086c0e57 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  33:        0x108680962 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  34:        0x1086c112b - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  35:        0x1086c0e57 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  36:        0x108680962 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  37:        0x1086c112b - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  38:        0x108680ac0 - rustc::hir::intravisit::walk_expr::h0397f9ae7f134d2a
  39:        0x1086c0e57 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::he608f8ca4d917271
  40:        0x1086bf328 - rustc_typeck::check::regionck::RegionCtxt::visit_fn_body::h9383541c740ad04f
  41:        0x10871339c - rustc_typeck::check::typeck_tables_of::{{closure}}::h9a71412a06e16e09
  42:        0x10870f506 - rustc_typeck::check::typeck_tables_of::h4c89c6362ba4db84
  43:        0x1096008ba - rustc::dep_graph::graph::DepGraph::with_task_impl::hfaaa0e8dd4a0683a
  44:        0x1098ea519 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::force::hecafdde183810b42
  45:        0x1098eb48f - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::try_get::hf32391a4fa1f99f3
  46:        0x1099f7230 - rustc::ty::maps::TyCtxtAt::typeck_tables_of::hee4782d933b10f61
  47:        0x1098ea170 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::ensure::ha9294646c2164cb8
  48:        0x10870e93f - rustc_typeck::check::typeck_item_bodies::hb8730a2b77d5113b
  49:        0x1095dd79c - rustc::dep_graph::graph::DepGraph::with_task_impl::h544a1fc76fc91d26
  50:        0x1098e8297 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_item_bodies<'tcx>>::force::hbf0e0dcc0f73b4d9
  51:        0x1098e90ff - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_item_bodies<'tcx>>::try_get::h9caadae2816555ce
  52:        0x1099f714b - rustc::ty::maps::TyCtxtAt::typeck_item_bodies::h32f9c1eb26d6a593
  53:        0x1099f3931 - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::typeck_item_bodies::h8ce9a5476cf40294
  54:        0x108761fe4 - rustc_typeck::check_crate::heab3b444e6275688
  55:        0x1080d7b53 - rustc::ty::context::TyCtxt::create_and_enter::h5f0cc76ebc317e07
  56:        0x108143404 - rustc_driver::driver::compile_input::h2094170bfafa48cd
  57:        0x10815fd26 - rustc_driver::run_compiler::hd4f50210430c210c
  58:        0x108087205 - std::sys_common::backtrace::__rust_begin_short_backtrace::h742c5f751224f94f
  59:        0x10ae6489e - __rust_maybe_catch_panic
  60:        0x1080c4935 - <F as alloc::boxed::FnBox<A>>::call_box::ha17229e1eb118d33
  61:        0x10ae5955b - std::sys::unix::thread::Thread::new::thread_start::heb1c5cf201612037
  62:     0x7fff57f8b6c0 - _pthread_body
  63:     0x7fff57f8b56c - _pthread_start
```
</details>

I get the same error message if I forget a "extern crate" or have a case-sensitive error in my "use XXX" statements. Its pretty easy to reproduce. See the following code. "Thread_rng" should be "thread_rng". Same happens if you miss the "extern crate rng"

extern crate rand;

use rand::{Rng,Thread_rng};

fn main() {
    println!("Hello, world! {}",*thread_rng().choose(&[0, 1, 2, 3]).unwrap());

}

## Output
   Compiling rngerr v0.1.0 (file:///D:/rustproj/rngerr)
     Running `rustc --crate-name rngerr src\main.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=90b29a1b474d9a34 -C extra-filename=-90b29a1b474d9a34 --out-dir D:\rustproj\rngerr\target\debug\deps -C incremental=D:\rustproj\rngerr\target\debug\incremental -L dependency=D:\rustproj\rngerr\target\debug\deps --extern rand=D:\rustproj\rngerr\target\debug\deps\librand-39fb58d4902b5813.rlib`
error[E0432]: unresolved import `rand::Thread_rng`
 --> src\main.rs:3:16
  |
3 | use rand::{Rng,Thread_rng};
  |                ^^^^^^^^^^ no `Thread_rng` in the root. Did you mean to use `thread_rng`?

error[E0425]: cannot find function `thread_rng` in this scope
 --> src\main.rs:6:34
  |
6 |     println!("Hello, world! {}",*thread_rng().choose(&[0, 1, 2, 3]).unwrap());
  |                                  ^^^^^^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
  |
3 | use rand::thread_rng;
  |

warning: unused import: `Thread_rng`
 --> src\main.rs:3:16
  |
3 | use rand::{Rng,Thread_rng};
  |                ^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default

error: internal compiler error: librustc\ich\impls_ty.rs:907: ty::TypeVariants::hash_stable() - Unexpected variant TyInfer(?0).

thread 'rustc' panicked at 'Box<Any>', librustc_errors\lib.rs:540:9
stack backtrace:
   0:     0x7ff9c9a361fa - _rdl_shrink_in_place
   1:     0x7ff9c9a39e52 - std::panicking::take_hook::h7e6bff1200b4d3f9
   2:     0x7ff9c9a39ab3 - std::panicking::take_hook::h7e6bff1200b4d3f9
   3:     0x7ff9c8614add - <unknown>
   4:     0x7ff9c9a3a456 - std::panicking::rust_panic_with_hook::h17a5bc6b5077376c
   5:     0x7ff9ce391a5e - <unknown>
   6:     0x7ff9ce3b5da8 - rustc_errors::Handler::bug::hb8b71f117d2421da
   7:     0x7ff9c88f45de - rustc::session::bug_fmt::h12a1661df948d163
   8:     0x7ff9c89c94cb - rustc::ty::context::tls::span_debug::h4d0feb810a227a29
   9:     0x7ff9c89c929e - rustc::ty::context::tls::span_debug::h4d0feb810a227a29
  10:     0x7ff9c86128fe - <unknown>
  11:     0x7ff9c8612859 - <unknown>
  12:     0x7ff9c89c9239 - rustc::ty::context::tls::span_debug::h4d0feb810a227a29
  13:     0x7ff9c89c9471 - rustc::ty::context::tls::span_debug::h4d0feb810a227a29
  14:     0x7ff9c88f44ec - rustc::session::bug_fmt::h12a1661df948d163
  15:     0x7ff9c88f4452 - rustc::session::bug_fmt::h12a1661df948d163
  16:     0x7ff9c88135d3 - rustc::ich::impls_hir::<impl rustc_data_structures::stable_hasher::ToStableHashKey<rustc::ich::hcx::StableHashingContext<'a>> for rustc::hir::TraitCandidate>::to_stable_hash_key::h9eed4c898167a695
  17:     0x7ff9c88131af - rustc::ich::impls_hir::<impl rustc_data_structures::stable_hasher::ToStableHashKey<rustc::ich::hcx::StableHashingContext<'a>> for rustc::hir::TraitCandidate>::to_stable_hash_key::h9eed4c898167a695
  18:     0x7ff9c89eaa2c - rustc::dep_graph::dep_node::DepNode::new::haaa5ab305ec2c44d
  19:     0x7ff9c8bd5f70 - rustc::ty::maps::<impl rustc::ty::maps::queries::dropck_outlives<'tcx>>::try_get::h213e7ce9ff55095c
  20:     0x7ff9c8bef4e6 - rustc::ty::maps::TyCtxtAt::dropck_outlives::h78ac1228027cebbb
  21:     0x7ff9c894f7f1 - rustc::traits::query::dropck_outlives::<impl rustc::infer::at::At<'cx, 'gcx, 'tcx>>::dropck_outlives::hd82e3d06092885ba
  22:     0x7ff9cdd1cc7c - <rustc_typeck::check::autoderef::Autoderef<'a, 'gcx, 'tcx> as core::iter::iterator::Iterator>::next::h9e3684d66782109d
  23:     0x7ff9cdd2af96 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  24:     0x7ff9cdd2887a - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  25:     0x7ff9cdd28eb7 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  26:     0x7ff9cdce557e - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  27:     0x7ff9cdd292d1 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  28:     0x7ff9cdce557e - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  29:     0x7ff9cdd292d1 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  30:     0x7ff9cdd29a3f - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  31:     0x7ff9cdd28eb7 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  32:     0x7ff9cdce57fe - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  33:     0x7ff9cdd29109 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  34:     0x7ff9cdce5725 - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  35:     0x7ff9cdd29453 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  36:     0x7ff9cdd28eb7 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  37:     0x7ff9cdce570e - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  38:     0x7ff9cdd29771 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  39:     0x7ff9cdce570e - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  40:     0x7ff9cdd29771 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  41:     0x7ff9cdce5900 - <rustc_typeck::outlives::test::OutlivesTest<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_impl_item::h5212b77bfc33afaf
  42:     0x7ff9cdd29109 - <rustc_typeck::check::regionck::RegionCtxt<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_expr::hcb39bebdbb74d460
  43:     0x7ff9cdd27306 - <rustc_typeck::check::writeback::Resolver<'cx, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_region::hb1622b89d3435e74
  44:     0x7ff9cdd869be - <rustc_typeck::check::CheckItemTypesVisitor<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_item::h540dd53913ae351e
  45:     0x7ff9cdd82bb6 - <rustc_typeck::check::CheckItemTypesVisitor<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_item::h540dd53913ae351e
  46:     0x7ff9c873fbba - rustc::dep_graph::graph::DepGraph::assert_ignored::h144efdc901c11500
  47:     0x7ff9c8aa6c76 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::ensure::h2b2d41c86f9a6e6e
  48:     0x7ff9c8aa7dad - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::try_get::h0bc6807c7004a3fd
  49:     0x7ff9c8be81df - rustc::ty::maps::TyCtxtAt::typeck_tables_of::hc49daa61101a1c23
  50:     0x7ff9c8aa6881 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_tables_of<'tcx>>::ensure::h2b2d41c86f9a6e6e
  51:     0x7ff9cdd821c2 - <rustc_typeck::check::CheckItemTypesVisitor<'a, 'tcx> as rustc::hir::itemlikevisit::ItemLikeVisitor<'tcx>>::visit_item::h540dd53913ae351e
  52:     0x7ff9c875a7dc - rustc::dep_graph::graph::DepGraph::assert_ignored::h144efdc901c11500
  53:     0x7ff9c8aa43d1 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_item_bodies<'tcx>>::ensure::ha12c0309982046bf
  54:     0x7ff9c8aa5444 - rustc::ty::maps::<impl rustc::ty::maps::queries::typeck_item_bodies<'tcx>>::try_get::hdf04a0cb4267327c
  55:     0x7ff9c8be80c7 - rustc::ty::maps::TyCtxtAt::typeck_item_bodies::h7155964f4ae16ed6
  56:     0x7ff9c8be4412 - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::typeck_item_bodies::hc8981e99eb7d74ce
  57:     0x7ff9cdde526e - rustc_typeck::check_crate::h82688740b0ec7d1c
  58:     0x7ff9da8fb6e7 - <env_logger::Logger as log::Log>::flush::he47d2badcc01211e
  59:     0x7ff9da972064 - rustc_driver::driver::compile_input::hd50b5e33b4a9e551
  60:     0x7ff9da997f80 - rustc_driver::run_compiler::h271536ce91a9abed
  61:     0x7ff9da894b60 - <unknown>
  62:     0x7ff9c9a4ae11 - _rust_maybe_catch_panic
  63:     0x7ff9da8d4f94 - <env_logger::Logger as log::Log>::flush::he47d2badcc01211e
  64:     0x7ff9c9a48b4e - std::sys::windows::thread::Thread::new::he6cb72092aa085f2
  65:     0x7ffa068a1fe3 - BaseThreadInitThunk
  66:     0x7ffa0942efc0 - RtlUserThreadStart

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.26.0-nightly (521d91c6b 2018-03-14) running on x86_64-pc-windows-msvc

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `rngerr`.

Caused by:
  process didn't exit successfully: `rustc --crate-name rngerr src\main.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 -C metadata=90b29a1b474d9a34 -C extra-filename=-90b29a1b474d9a34 --out-dir D:\rustproj\rngerr\target\debug\deps -C incremental=D:\rustproj\rngerr\target\debug\incremental -L dependency=D:\rustproj\rngerr\target\debug\deps --extern rand=D:\rustproj\rngerr\target\debug\deps\librand-39fb58d4902b5813.rlib` (exit code: 101)

Process finished with exit code 101

--- a/src/librustc/ich/impls_ty.rs
+++ b/src/librustc/ich/impls_ty.rs
@@ -902,10 +902,56 @@ for ty::TypeVariants<'gcx>
             TyForeign(def_id) => {
                 def_id.hash_stable(hcx, hasher);
             }
-            TyInfer(..) => {
-                bug!("ty::TypeVariants::hash_stable() - Unexpected variant {:?}.", *self)
+            TyInfer(infer_ty) => {
+                infer_ty.hash_stable(hcx, hasher);
             }
         }
+    }
+}
+
+impl_stable_hash_for!(enum ty::InferTy {
+    TyVar(a),
+    IntVar(a),
+    FloatVar(a),
+    FreshTy(a),
+    FreshIntTy(a),
+    FreshFloatTy(a),
+    CanonicalTy(a),
+});
+
+impl<'a, 'gcx> HashStable<StableHashingContext<'a>>
+for ty::TyVid
+{
+    fn hash_stable<W: StableHasherResult>(&self,
+                                          _hcx: &mut StableHashingContext<'a>,
+                                          _hasher: &mut StableHasher<W>) {
+        // TyVid values are confined to an inference context and hence
+        // should not be hashed.
+        bug!("ty::TypeVariants::hash_stable() - can't hash a TyVid {:?}.", *self)
+    }
+}
+
+impl<'a, 'gcx> HashStable<StableHashingContext<'a>>
+for ty::IntVid
+{
+    fn hash_stable<W: StableHasherResult>(&self,
+                                          _hcx: &mut StableHashingContext<'a>,
+                                          _hasher: &mut StableHasher<W>) {
+        // IntVid values are confined to an inference context and hence
+        // should not be hashed.
+        bug!("ty::TypeVariants::hash_stable() - can't hash an IntVid {:?}.", *self)
+    }
+}
+
+impl<'a, 'gcx> HashStable<StableHashingContext<'a>>
+for ty::FloatVid
+{
+    fn hash_stable<W: StableHasherResult>(&self,
+                                          _hcx: &mut StableHashingContext<'a>,
+                                          _hasher: &mut StableHasher<W>) {
+        // FloatVid values are confined to an inference context and hence
+        // should not be hashed.
+        bug!("ty::TypeVariants::hash_stable() - can't hash a FloatVid {:?}.", *self)
     }
 }
 

--- a/src/librustc/macros.rs
+++ b/src/librustc/macros.rs
@@ -72,7 +72,7 @@ macro_rules! __impl_stable_hash_field {
 
 #[macro_export]
 macro_rules! impl_stable_hash_for {
-    (enum $enum_name:path { $( $variant:ident $( ( $($arg:ident),* ) )* ),* }) => {
+    (enum $enum_name:path { $( $variant:ident $( ( $($arg:ident),* ) )* ),* $(,)* }) => {
         impl<'a, 'tcx> ::rustc_data_structures::stable_hasher::HashStable<$crate::ich::StableHashingContext<'a>> for $enum_name {
             #[inline]
             fn hash_stable<W: ::rustc_data_structures::stable_hasher::StableHasherResult>(&self,

--- a/src/librustc/traits/query/dropck_outlives.rs
+++ b/src/librustc/traits/query/dropck_outlives.rs
@@ -51,6 +51,7 @@ impl<'cx, 'gcx, 'tcx> At<'cx, 'gcx, 'tcx> {
         let gcx = tcx.global_tcx();
         let (c_ty, orig_values) = self.infcx.canonicalize_query(&self.param_env.and(ty));
         let span = self.cause.span;
+        debug!("c_ty = {:?}", c_ty);
         match &gcx.dropck_outlives(c_ty) {
             Ok(result) if result.is_proven() => {
                 match self.infcx.instantiate_query_result(

--- a/src/test/incremental/issue-49043.rs
+++ b/src/test/incremental/issue-49043.rs
@@ -1,0 +1,22 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Regression test for hashing involving canonical variables.  In this
+// test -- which has an intensional error -- the type of the value
+// being dropped winds up including a type variable. Canonicalization
+// would then produce a `?0` which -- in turn -- triggered an ICE in
+// hashing.
+
+// revisions:cfail1
+
+fn main() {
+    println!("Hello, world! {}",*thread_rng().choose(&[0, 1, 2, 3]).unwrap());
+    //[cfail1]~^ ERROR cannot find function `thread_rng`
+}

